"Microbatch" incremental strategy (#6194)
fishtown-analytics/docs.getdbt.com: website/docs/docs/build/incremental-microbatch.md

---
title: "About microbatch incremental models"
description: "Learn about the 'microbatch' strategy for incremental models."
id: "incremental-microbatch"
---

> 

Reviewer: ```suggestion

# About microbatch incremental models <Lifecycle status="beta" />
```

---
title: "About microbatch incremental models"
description: "Learn about the 'microbatch' strategy for incremental models."
id: "incremental-microbatch"
---

# About microbatch incremental models <Lifecycle status="beta" />

> :::info Microbatch <Lifecycle status="beta" />

Reviewer: ```suggestion

:::info Microbatch
```

---
title: "About microbatch incremental models"
description: "Learn about the 'microbatch' strategy for incremental models."
id: "incremental-microbatch"
---

# About microbatch incremental models <Lifecycle status="beta" />

> :::info Microbatch

Reviewer 2: @jtcohen6 Thoughts on calling out that this _feature_ is in beta - as I anticipate there will be some changes ahead of our 1.9 final release based on community feedback

Reviewer: added beta in there as Jerco originally added it and i removed it since i added in the h1! sorry!

The `microbatch` strategy is available in [dbt Cloud Versionless](/docs/dbt-versions/upgrade-dbt-version-in-cloud#versionless) and dbt Core v1.9.

> 

Reviewer: do we mean to set it to true in their dbt cloud environments, yes?

Reviewer 2: I think we can just say "to true" and end there

Read and participate in the discussion: [dbt-core#10672](https://github.com/dbt-labs/dbt-core/discussions/10672)

:::

## What is microbatch?

> Incremental models in dbt are a [materialization](https://docs.getdbt.com/docs/build/materializations) designed to efficiently update your data warehouse tables by only transforming and loading _new or changed data_ since the last run. Instead of reprocessing an entire dataset every time, incremental models append, update, or replace rows in the existing table with the new data just processed. This can significantly reduce the time and resources required for your data transformations.

Reviewer: ```suggestion
Incremental models in dbt are a [materialization](/docs/build/materializations) designed to efficiently update your data warehouse tables by only transforming and loading _new or changed data_ since the last run. Instead of reprocessing an entire dataset every time, incremental models append, update, or replace rows in the existing table with the new data just processed. This can significantly reduce the time and resources required for your data transformations.
```

> Microbatch incremental models make it possible to process transformations on large time-series datasets with efficiency and resiliency. When dbt runs a microbatch model — whether for the first time, during incremental runs, or during manual backfills — it will split the processing into multiple queries (or "batches"), based on the `event_time` column you configure.

Reviewer: ```suggestion
Microbatch incremental models make it possible to process transformations on large time-series datasets with efficiency and resiliency. When dbt runs a microbatch model — whether for the first time, during incremental runs, or during manual backfills — it will split the processing into multiple queries (or "batches"), based on the `event_time` column you configure. Instead of processing all of your data at once. Since each "batch" is based on a time period, like a single day, it makes it much faster and more efficient to update large datasets, especially when you're working with data that changes over time (like new records being added daily).
```

Author: thanks for the suggestion! I will add more here

> Microbatch incremental models make it possible to process transformations on large time-series datasets with efficiency and resiliency. When dbt runs a microbatch model — whether for the first time, during incremental runs, or during manual backfills — it will split the processing into multiple queries (or "batches"), based on the `event_time` column you configure.

Reviewer 2: ```suggestion
Microbatch incremental models make it possible to process transformations on large time-series datasets with efficiency and resiliency. When dbt runs a microbatch model — whether for the first time, during incremental runs, or during manual backfills — it will split the processing into multiple queries (or "batches"), based on the `event_time` columns and `batch_size` you configure.
```

> Where other incremental strategies operate only on "old" and "new" data, microbatch models treat each "batch" of data as a unit that can be built or replaced on its own. Each batch is independent and <Term id="idempotent" />. This is a powerful abstraction that makes it possible for dbt to run batches separately — in the future, concurrently — and to retry them independently.

Reviewer 2: I know we have an example in the "## How does `microbatch` compare to other incremental strategies?" section - but maybe we should have one here as well?

Author: good all! adding an example

### Available configs

> - `event_time` - The column indicating "at what time did the row occur" (for both your microbatch model and its direct parents)

Reviewer 2: maybe a callout that `event_time` and `begin` need to be in UTC? (also true of the CLI args `--event-time-start/end`

Reviewer 2: or maybe having the "timezone"s section is better!

    
<Lightbox src="/img/docs/building-a-dbt-project/microbatch/event_time.png" title="The event_time column configures the real-world time of this record"/>
    

> - `batch_size` (string, optional) - The granularity of your batches. The default is `day`, and currently that is the only granularity supported.

Reviewer 2: Thoughts on making this a table instead of a bulleted list?

Reviewer: adding table

- `lookback` (integer, optional) - Process X batches prior to the latest bookmark, in order to capture late-arriving records. The default value is `0`.

> - `begin` (date, optional) - The "beginning of time" for your data. This is the starting point for any initial or full-refresh builds. For example, a daily-grain microbatch model run on `2024-10-01` with `begin = '2023-10-01` will process 366 batches. (It's a leap year!)

Reviewer 2: I don't _think_ this is optional

As a best practice, we recommend configuring `full_refresh: False` on microbatch models so that they ignore invocations with the `--full-refresh` flag. If you need to reprocess historical data, do so with a targeted backfill.

### Usage

**You write your model query to process (read and return) one day of data**. You don’t need to think about `is_incremental` filtering or DML (upserting/merging/replacing) - we take care of that for you.

dbt will then evaluate which batches need to be loaded, break them up into a SQL query per batch, and load each one independently.

dbt will automatically filter upstream inputs (`source` or `ref`) that define `event_time`, based on the `lookback` and `batch_size` configs for this model.

> During standard incremental runs, dbt will process new batches and any according to the configured `lookback` (with one query per batch)

Reviewer 2: ```suggestion
During standard incremental runs, dbt will process new batches and any according to the current timestamp and configured `lookback` (with one query per batch).
```

> <Lightbox src="/img/docs/building-a-dbt-project/microbatch/microbatch_lookback.png" title="Configure a lookback to reprocess additional batches during standard incremental runs"/>

Reviewer: will fix img

> If there’s an upstream model that configures `event_time`, but you *don’t* want the reference to it to be filtered, you can specify `ref('upstream_model').render()` to opt-out of auto-filtering.

Reviewer 2: Thoughts on nesting this under the sentence "dbt will automatically filter upstream inputs (`source` or `ref`) that define `event_time`, based on the `lookback` and `batch_size` configs for this model."

Author: I tried both above / below. I'd rather stick this below the image, because it feels like an exception rather than the rule — but I could be wrong! If we expect people to need to do this very often, that's one more vote in favor of a more intuitive (less-ugly) syntax

Backfilling a microbatch model is as simple as selecting it to run or build, and specifying a "start" and "end" for `event_time`. As always, dbt will process the batches between the start and end as independent queries.

```bash
dbt run --event-time-start "2024-09-01" --event-time-end "2024-09-04"
```

<Lightbox src="/img/docs/building-a-dbt-project/microbatch/microbatch_backfill.png" title="Configure a lookback to reprocess additional batches during standard incremental runs"/>

### Retry

If one or more of your batches fail, you can use `dbt retry` to reprocess _only_ the failed batches.

![Partial retry](https://github.com/user-attachments/assets/f94c4797-dcc7-4875-9623-639f70c97b8f)

### Timezones

For now, dbt assumes that all values supplied are in UTC:

- `event_time`
- `begin`
- `--event-time-start` and `--event-time-end`

While we may consider adding support for custom timezones in the future, we also believe that defining these values in UTC makes everyone's lives easier.

## How does `microbatch` compare to other incremental strategies?

> Most incremental models rely on the end user (you) to explicitly tell dbt what "new" means, in the context of each model, by writing a filter in an `is_incremental()` block. You are responsibly for crafting this SQL in a way that queries `this` to check when the most recent record was last loaded, with an optional look-back window for late-arriving records. Other incremental strategies will control _how_ the data is being added into the table — whether append-only `insert`, `delete` + `insert`, `merge`, `insert overwrite`, etc — but they all have this in common.

Reviewer 2: "Other incremental strategies will control _how_ the data is being added into the table"

maybe add a callout of how microbatch choose which of these "how"s is best?

For this incremental model:

- "New" records are those with a `date_day` greater than the maximum `date_day` that has previously been loaded
- The lookback window is 3 days
- When there are new records for a given `date_day`, the existing data for `date_day` is deleted and the new data is inserted

Let’s take our same example from before, and instead use the new `microbatch` incremental strategy:

```sql
{{
    config(
        materialized='incremental',
        incremental_strategy='microbatch',
        event_time='event_occured_at',
        batch_size='day',
        lookback=3,
        begin='2020-01-01',
        full_refresh=false
    )
}}

select * from {{ ref('stg_events') }} # this ref will be auto-filtered
```

Where you’ve also set an `event_time` for the model’s direct parents - in this case `stg_events`:

```yaml
models:
  - name: stg_events
     config:
       event_time: my_time_field
```

And that’s it! When you run the model, each batch templates a separate query. The batch for `2024-10-01` would template:

```sql
select * from (
    select * from {{ ref('stg_events') }}
    where my_time_field >= '2024-10-01 00:00:00'
      and my_time_field < '2024-10-02 00:00:00'

> ) # this ref will be auto-filtered

Reviewer 2: ```suggestion
) 
```

---
title: "About microbatch incremental models"
description: "Learn about the 'microbatch' strategy for incremental models."
id: "incremental-microbatch"
---

# About microbatch incremental models <Lifecycle status="beta" />

:::info Microbatch

> The `microbatch` strategy is available in beta for [dbt Cloud Versionless](/docs/dbt-versions/upgrade-dbt-version-in-cloud#versionless) and dbt Core v1.9. To enable this feature,  set `DBT_EXPERIMENTAL_MICROBATCH` to true in your project.

Reviewer: ```suggestion
The `microbatch` strategy is available in beta for [dbt Cloud Versionless](/docs/dbt-versions/upgrade-dbt-version-in-cloud#versionless) and dbt Core v1.9. To enable this feature,  set the environment variable `DBT_EXPERIMENTAL_MICROBATCH` to true in your project.
```

> The `microbatch` strategy is available in beta for [dbt Cloud Versionless](/docs/dbt-versions/upgrade-dbt-version-in-cloud#versionless) and dbt Core v1.9. To enable this feature,  set the environment variable `DBT_EXPERIMENTAL_MICROBATCH` to true in your project.

Reviewer 2: ```suggestion
The `microbatch` strategy is available in beta for [dbt Cloud Versionless](/docs/dbt-versions/upgrade-dbt-version-in-cloud#versionless) and dbt Core v1.9. To enable this feature,  set the environment variable `DBT_EXPERIMENTAL_MICROBATCH` to true.
```

### Usage

**You write your model query to process (read and return) one day of data**. You don’t need to think about `is_incremental` filtering or DML (upserting/merging/replacing) - we take care of that for you.

dbt will then evaluate which batches need to be loaded, break them up into a SQL query per batch, and load each one independently.

dbt will automatically filter upstream inputs (`source` or `ref`) that define `event_time`, based on the `lookback` and `batch_size` configs for this model.

During standard incremental runs, dbt will process new batches and any according to the configured `lookback` (with one query per batch)
    
<Lightbox src="/img/docs/building-a-dbt-project/microbatch/microbatch_lookback.png" title="Configure a lookback to reprocess additional batches during standard incremental runs"/>

If there’s an upstream model that configures `event_time`, but you *don’t* want the reference to it to be filtered, you can specify `ref('upstream_model').render()` to opt-out of auto-filtering.

> dbt will evaluate which batches need to be loaded **by processing the current batch (current_timestamp) + any batches in your configured lookback**, break them up into a SQL query per batch, and load them all independently. 

Author: redundant

        where date_day >= (select {{ dbt.dateadd("day", -3, "max(date_day)") }} from {{ this }})  
    {% endif %}

```

For this incremental model:

- "New" records are those with a `date_day` greater than the maximum `date_day` that has previously been loaded
- The lookback window is 3 days
- When there are new records for a given `date_day`, the existing data for `date_day` is deleted and the new data is inserted

Let’s take our same example from before, and instead use the new `microbatch` incremental strategy:

```sql
{{
    config(
        materialized='incremental',
        incremental_strategy='microbatch',
        event_time='event_occured_at',
        batch_size='day',
        lookback=3,
        begin='2020-01-01',
        full_refresh=false
    )
}}

select * from {{ ref('stg_events') }} # this ref will be auto-filtered
```

Where you’ve also set an `event_time` for the model’s direct parents - in this case `stg_events`:

```yaml
models:
  - name: stg_events
     config:
       event_time: my_time_field
```

And that’s it! When you run the model, each batch templates a separate query. The batch for `2024-10-01` would template:

```sql
select * from (

>     select * from {{ ref('stg_events') }}

Reviewer 2: should we show the compiled SQL for the ref statement here?

Author: yes... obviously

    )

),

customers as (

    select * from "analytics"."customers"

),

...
```

</File>

</TabItem>

</Tabs>

dbt will instruct the data platform to take the result of each batch query and insert, update, or replace the contents of the `analytics.sessions` table for the same day of data. To perform this operation, dbt will use the most efficient atomic mechanism for "full batch" replacement that is available on each data platform.

It does not matter whether the table already contains data for that day, or not. Given the same input data, no matter how many times a batch is reprocessed, the resulting table is the same.

<Lightbox src="/img/docs/building-a-dbt-project/microbatch/microbatch_filters.png" title="The event_time column configures the real-world time of this record"/>

### Relevant configs

Several configurations are relevant to microbatch models, and some are required:

| Config   | Type | Description   | Default |
|----------|------|---------------|---------|
| `event_time` | Column             | The column indicating "at what time did the row occur." Required for your microbatch model and any direct parents that should be filtered.          | N/A     |

> | `begin`      | Date    | The "beginning of time" for the microbatch model. This is the starting point for any initial or full-refresh builds. For example, a daily-grain microbatch model run on `2024-10-01` with `begin = '2023-10-01` will process 366 batches (it's a leap year!) plus the batch for "today."        | N/A     |

Reviewer: assuming `begin` is required , right?
